### PR TITLE
P10 support

### PR DESCRIPTION
**Pull request by npanpaliya**
## Checklist before submitting

- [ ] Did you read the [contributor guide](https://github.com/open-ce/open-ce/blob/main/CONTRIBUTING.md)?
- [ ] Did you update any affected [documentation](https://github.com/open-ce/open-ce/blob/main/doc/)?
- [ ] Did you write any tests to validate this change?

## Description

[Fixes # (issue).](https://github.com/open-ce/open-ce/issues/585)

## Review process to land 

1. All tests and other checks must succeed.
2. At least one [maintainer](https://github.com/open-ce/open-ce/blob/main/MAINTAINERS.md) must review and approve.
3. If any  [maintainer](https://github.com/open-ce/open-ce/blob/main/MAINTAINERS.md) requests changes, they must be addressed.

### CODE REVIEW

**Review by alexlang74 — COMMENTED** [review]
very good and comprehensive. Where should we add the requirement to certain RHEL 8.5 runtime libraries for Pytorch and sentencepiece?

File: doc/README.open_ce_build.md [hunk 1]
```diff
@@ -172,6 +173,8 @@ optional arguments:
   --conda_pkg_format CONDA_PKG_FORMAT
                         Conda package format to be used, such as "tarball" or
                         "conda". (default: conda)
+  --ppc_arch PPC_ARCH   Power Architecture to build for, such as "p9" or "p10". Use p10 for power10 enabled builds.
```
Comment (alexlang74) [inline doc/README.open_ce_build.md:176]: > such as "p9" or "p10"

I suggest to write the actual supported values. is it p9, p10, or also p8 ?

File: doc/README.open_ce_build.md [hunk 2]
```diff
@@ -410,3 +416,37 @@ The Dockerfile for this runtime image is located in [`images/opence-runtime/Dock
...
+
+### Building packages with Power10 MMA optimization
+
+#### System Requirement
+System: RHEL 8.5 or above
+OS: Linux
+Power Architecture: Power9/Power10
+GCC Compiler: GCC10 and GCC11
+
+One can build Power10 enabled packages on Power9 or Power10 with above system requirements. To install GCC 10 or 11, following command can be used -
```
Comment (alexlang74) [inline doc/README.open_ce_build.md:428]: > One can build Power10 enabled packages on Power9 or Power10 with above system requirements.

I suggest to highlight this further:  
One can build Power10 enabled packages with above system requirements. Note that Power10  is _not_ required on your build system. The libraries can be built on Power9 as well.

File: doc/README.open_ce_build.md [hunk 3]
```diff
@@ -410,3 +416,37 @@ The Dockerfile for this runtime image is located in [`images/opence-runtime/Dock
...
+
+#### Build packages
+Power10 MMA Optimization is applicable for cpu only builds. One has to use `--ppc_arch=p10` flag in the `open-ce build env` or `open-ce build feedstock` command to build P10 enabled packages. Another important argument which is must to build these packages is `--conda_build_config=open-ce/envs/conda_build_config.yaml,open-ce/envs/conda_build_config_p10.yaml`.[`conda_build_config_p10.yaml`](https://github.com/open-ce/open-ce/blob/main/envs/conda_build_config_p10.yaml) contains Power10 specific settings.
+
+For example:
+```shell
+    open-ce build env --build_type=cpu --ppc_arch=p10 --conda_build_config=open-ce/envs/conda_build_config.yaml,open-ce/envs/conda_build_config_p10.yaml tensorflow-env.yaml
+```
+
+Packages built with above commands runs on Power10 with MMA optimization and on Power9 without MMA optimization. GCC10 is also needed at runtime for some of these packages to work, may it be P9 or P10.
```
Comment (alexlang74) [inline doc/README.open_ce_build.md:452]: > Packages built with above commands runs

Packages built with above commands run~s~

I suggest to put _GCC10 is also needed at runtime for some of these packages to work, may it be P9 or P10._ into its own section, like so

### Using packages with Power10 MMA optimization

When using packages that were built with `ppc_arch=p10`, note that:
- These packages will work on Power9 or Power10, but not on Power8
- When using `PyTorch` or `sentencepiece`, gcc10 needs to be present on the system. These two packages require some gcc10 libraries to be present at runtime. This applies to both Power9 and Power10 systems.

**Review by alexlang74 — APPROVED** [review]
Looks good, please see my comments

File: README.md [hunk 4]
```diff
@@ -148,6 +148,11 @@ After performing the build using the `open-ce build env` tool, the `open-ce test
 
 After performing the build using `open-ce build env`, the `open-ce build image` command can be used to create a runtime container image containing the newly created conda channel, as well as a conda environment with the newly build Open-CE packages. For more details on `open-ce build image`, please see [`doc/README.open_ce_build.md`](doc/README.open_ce_build.md#open-ce-build-image-sub-command).
 
+### Building Packages with Power10 MMA Optimization
+
+One can build the major Open-CE libraries like Tensorflow, Pytorch, Xgboost, etc. with Power10 MMA optimization.
```
Comment (cdeepali) [inline README.md:153]:
```suggestion
One can build the major Open-CE libraries like TensorFlow, PyTorch, XGBoost, etc. with Power10 MMA optimization.
```

File: doc/README.open_ce_build.md [hunk 5]
```diff
@@ -410,3 +416,37 @@ The Dockerfile for this runtime image is located in [`images/opence-runtime/Dock
 This file can also be built and run manually and supports GPUs if the system is set up with
 `nvidia-container-runtime`. When building the Dockerfile directly, it does require a few arguents,
 check the Dockerfile for details.
+
+### Building packages with Power10 MMA optimization
+
+#### System Requirement
+System: RHEL 8.5 or above
```
Comment (cdeepali) [inline doc/README.open_ce_build.md:423]: The requirements are not appearing as list .md. Slight correction needed here.

File: doc/README.open_ce_build.md (hunk 1 above)
Comment (npanpaliya) [reply doc/README.open_ce_build.md:176]: Actually in code, it's either p9 or p10. "p9" means "p8" as well as "p9" type of builds (our type1 builds).

File: doc/README.open_ce_build.md (hunk 2 above)
Comment (npanpaliya) [reply doc/README.open_ce_build.md:428]: Sounds good!

File: doc/README.open_ce_build.md (hunk 1 above)
Comment (alexlang74) [reply doc/README.open_ce_build.md:176]: Ok, then it should say. 
Power Architecture to build for. Values: `p9` or `p10`.  `p9`: Libraries can be used on Power8, Power9 and Power 10, but do not use MMA acceleration. `p10`: Libraries can be used on Power9 and Power10, and use MMA acceleration on Power10.

**Review by alexlang74 — APPROVED** [review]

**Comment by npanpaliya** [comment]
The test coverage is decreased because I've not added any new tests for newly added code. We'd decided to add new tests during 1.6.0 work.

**Review by alexlang74 — APPROVED** [review]

File: open_ce/errors.py [hunk 6]
```diff
@@ -89,6 +89,9 @@ class Error(Enum):
     TEMP_BUILD_IMAGE_FILES = (37, "Error removing temporary files created during build image.")
     UNABLE_DOWNLOAD_SOURCE = (38, "Unable to download source for '{}'.")
     UNABLE_CLONE_SOURCE = (39, "Unable to clone source for '{}'.")
+    GCC10_11_COMPILER_NOT_FOUND = (40, "GCC10 and 11 are either not installed or are installed at non-default location." +
```
Comment (cdeepali) [inline open_ce/errors.py:92]: I think we can split this into two error codes.

File: doc/README.open_ce_build.md [hunk 7]
```diff
@@ -410,3 +422,42 @@ The Dockerfile for this runtime image is located in [`images/opence-runtime/Dock
 This file can also be built and run manually and supports GPUs if the system is set up with
 `nvidia-container-runtime`. When building the Dockerfile directly, it does require a few arguents,
 check the Dockerfile for details.
+
+### Building packages with Power10 MMA optimization
+
+#### System Requirement
```
Comment (cdeepali) [inline doc/README.open_ce_build.md:428]:
```suggestion
#### System Requirements
```

File: doc/README.open_ce_build.md [hunk 8]
```diff
@@ -410,3 +422,42 @@ The Dockerfile for this runtime image is located in [`images/opence-runtime/Dock
...
+For example:
+```shell
+    open-ce build env --build_type=cpu --ppc_arch=p10 --conda_build_config=open-ce/envs/conda_build_config.yaml,open-ce/envs/conda_build_config_p10.yaml tensorflow-env.yaml
+```
+
+#### Using packages with Power10 MMA optimization
+When using packages that were built with ppc_arch=p10, note that:
+
+* These packages will work on Power9 or Power10, but not on Power8
+* At runtime, gcc10 needs to be present on the system. Packages like TF, pytorch, sentencepiece, onnxruntime, etc. require some gcc10 libraries to be present at runtime. This applies to both Power9 and Power10 systems.
```
Comment (cdeepali) [inline doc/README.open_ce_build.md:463]:
```suggestion
* At runtime, GCC10 needs to be present on the system. Packages like TF, PyTorch, SentencePiece, ONNX Runtime, etc. require some GCC10 libraries to be present at runtime. This applies to both Power9 and Power10 systems.
```

File: open_ce/inputs.py [hunk 9]
```diff
@@ -363,6 +373,26 @@ def _create_env_config_paths(args):
...
+            if os.path.exists(os.environ["GCC_10_HOME"]):
+                PATH = os.environ["PATH"]
+                os.environ["PATH"] = "{0}:{1}".format(os.path.join(os.environ["GCC_10_HOME"], "bin"), PATH)
+                print("Path variable set to : ", os.environ["PATH"])
+            else:
+                raise OpenCEError(Error.GCC10_11_COMPILER_NOT_FOUND)
+            if "GCC_11_HOME" not in os.environ and os.path.exists(utils.DEFAULT_GCC_11_HOME_DIR):
+                os.environ["GCC_11_HOME"] = utils.DEFAULT_GCC_11_HOME_DIR
+            else:
+                raise OpenCEError(Error.GCC10_11_COMPILER_NOT_FOUND)
```
Comment (cdeepali) [inline open_ce/inputs.py:394]: I think this needs to be split into two steps like for GCC_10_HOME, as presently the else block gets called when GCC_11_HOME is defined.

Comment (npanpaliya) [reply open_ce/inputs.py:394]: Yes, thanks!

File: open_ce/errors.py (hunk 6 above)
Comment (npanpaliya) [reply open_ce/errors.py:92]: I think we can reword this error message but there is no need to have two separate error codes for GCC 10 and GCC 11. I'll reword the message.

Comment (cdeepali) [reply open_ce/errors.py]: Okay, thats fine too.

**Pull request merged**